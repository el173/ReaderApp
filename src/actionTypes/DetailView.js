export const GET_ITEM_DETAILS = 'GET_ITEM_DETAILS';
export const GET_ITEM_DETAILS_OK = 'GET_ITEM_DETAILS_OK';
