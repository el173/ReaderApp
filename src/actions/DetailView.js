import {takeLeading, put} from 'redux-saga/effects';

import {getItemDetails} from './Common';

import {
  GET_ITEM_DETAILS,
  GET_ITEM_DETAILS_OK,
  CLEAR_ITEM_DETAILS,
  CLEAR_ITEM_DETAILS_OK,
} from '../actionTypes';

const getItemInfo = function* ({payload}) {
  let commentsList = [];
  yield Promise.all(
    payload.map(async storyItem => {
      const itemDetails = await getItemDetails(storyItem);
      commentsList.push(itemDetails);
    }),
  );
  yield put({type: GET_ITEM_DETAILS_OK, payload: commentsList});
};

export function* getItemDetailsSaga() {
  yield takeLeading(GET_ITEM_DETAILS, getItemInfo);
}

const clearItemDetails = function* ({payload}) {
  yield put({type: CLEAR_ITEM_DETAILS_OK, payload: []});
};

export function* clearItemDetailsSaga() {
  yield takeLeading(CLEAR_ITEM_DETAILS, clearItemDetails);
}
