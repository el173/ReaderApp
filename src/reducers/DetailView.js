import {GET_ITEM_DETAILS_OK, CLEAR_ITEM_DETAILS_OK} from '../actionTypes';

const INITIAL_STATE = {
  itemDetails: [],
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case CLEAR_ITEM_DETAILS_OK:
    case GET_ITEM_DETAILS_OK:
      return {...state, itemDetails: action.payload};
    default:
      return state;
  }
};
