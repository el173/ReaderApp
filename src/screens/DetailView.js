import React, {useEffect} from 'react';
import {connect} from 'react-redux';
import {
  StyleSheet,
  View,
  ActivityIndicator,
  FlatList,
  BackHandler,
} from 'react-native';

import {CommentCard} from '../components';

import {GET_ITEM_DETAILS, CLEAR_ITEM_DETAILS} from '../actionTypes';

const DetailView = ({
  navigation,
  kids,
  getItemDetails,
  itemDetails,
  clearItemDetails,
}) => {
  const onBackPress = () => {
    clearItemDetails();
  };

  useEffect(() => {
    getItemDetails(kids);
    return onBackPress;
  }, []);

  useEffect(() => {
    BackHandler.addEventListener('hardwareBackPress', onBackPress);
    return () =>
      BackHandler.removeEventListener('hardwareBackPress', onBackPress);
  }, []);

  if (itemDetails.length <= 0) {
    return (
      <View style={styles.spinner}>
        <ActivityIndicator size="large" color={'#178DE7'} />
      </View>
    );
  }

  const keyExtractor = (item, index) => index;
  const renderItem = ({item, index}) => (
    <CommentCard
      key={`${index}-${item.time}`}
      newsItem={item}
      navigation={navigation}
    />
  );

  const ITEM_HEIGHT = 65;
  const getItemLayout = (data, index) => {
    return {
      length: ITEM_HEIGHT,
      offset: ITEM_HEIGHT * data.length,
      index,
    };
  };

  return (
    <View style={styles.container}>
      <FlatList
        data={itemDetails}
        keyExtractor={keyExtractor}
        renderItem={renderItem}
        getItemLayout={getItemLayout}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  spinner: {
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1,
  },
});

const mapStateToProps = (state, props) => {
  return {
    itemDetails: state.detailsViewReducer.itemDetails,
    ...props.route.params,
  };
};

function mapDispatchToProps(dispatch) {
  return {
    getItemDetails: payload => {
      dispatch({
        type: GET_ITEM_DETAILS,
        payload,
      });
    },
    clearItemDetails: payload => {
      dispatch({
        type: CLEAR_ITEM_DETAILS,
      });
    },
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(DetailView);
