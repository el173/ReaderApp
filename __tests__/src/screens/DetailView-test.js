import 'react-native';
import renderer from 'react-test-renderer';
import React from 'react';
import configureStore from 'redux-mock-store';
import {render, cleanup} from 'react-native-testing-library';
import {Provider} from 'react-redux';

import DetailView from '../../../src/screens/DetailView';

afterEach(() => {
  cleanup();
});
describe('Details View', () => {
  const mockStore = configureStore([]);

  const route = {params: {}};

  it('renders correctly', async () => {
    async () => {
      renderer.create(<DetailView />);
    };
  });

  it('renders no with data', () => {
    const detailsViewReducer = {
      itemDetails: [],
    };
    const store = mockStore({detailsViewReducer});
    const noDataScreen = render(
      <Provider store={store}>
        <DetailView kids={[]} route={route} />
      </Provider>,
    ).toJSON();
    expect(noDataScreen).toMatchSnapshot();
  });

  it('renders with data', () => {
    const detailsViewReducer = {
      itemDetails: [
        {
          by: 'svnpenn',
          id: 31387601,
          kids: [31387632],
          parent: 31387312,
          text: '&gt; func (c *container&gt;[S]) size() int {<p>This is not valid code.',
          time: 1652623124,
          type: 'comment',
        },
      ],
    };
    const store = mockStore({detailsViewReducer});
    const withData = render(
      <Provider store={store}>
        <DetailView kids={[31387601]} route={route} />
      </Provider>,
    ).toJSON();
    expect(withData).toMatchSnapshot();
  });
});
