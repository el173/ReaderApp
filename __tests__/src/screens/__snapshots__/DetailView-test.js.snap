// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details View renders no with data 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    color="#178DE7"
    size="large"
  />
</View>
`;

exports[`Details View renders with data 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "by": "svnpenn",
          "id": 31387601,
          "kids": Array [
            31387632,
          ],
          "parent": 31387312,
          "text": "&gt; func (c *container&gt;[S]) size() int {<p>This is not valid code.",
          "time": 1652623124,
          "type": "comment",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 10,
              "elevation": 3,
              "flex": 1,
              "marginBottom": 10,
              "marginHorizontal": 5,
              "padding": 15,
              "shadowColor": "rgba(0, 0, 0, 1.0)",
              "shadowOffset": Object {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.54,
              "shadowRadius": 2,
            }
          }
        >
          <View
            style={Object {}}
          >
            <View
              style={
                Array [
                  Object {},
                  undefined,
                  undefined,
                ]
              }
              testID="html"
            >
              <View
                style={
                  Array [
                    Object {},
                    undefined,
                    undefined,
                  ]
                }
                testID="body"
              >
                <Text
                  allowFontScaling={true}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "fontSize": 14,
                      },
                      undefined,
                      undefined,
                    ]
                  }
                >
                  &gt; func (c *container&gt;[S]) size() int {
                </Text>
                <View
                  style={
                    Array [
                      Object {
                        "marginBottom": 14,
                        "marginTop": 14,
                      },
                      undefined,
                      undefined,
                    ]
                  }
                  testID="p"
                >
                  <Text
                    allowFontScaling={true}
                    selectable={false}
                    style={
                      Array [
                        Object {
                          "fontSize": 14,
                        },
                        undefined,
                        undefined,
                      ]
                    }
                  >
                    This is not valid code.
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "marginTop": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 0.75,
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontSize": 12,
                  }
                }
              >
                Sun, 15 May 2022 13:58:44 GMT
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 0.25,
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontSize": 12,
                  }
                }
              >
                By: svnpenn
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
